### PR TITLE
Improved drawing in plot_rag_draw.py example

### PR DESCRIPTION
**Pull request by vighneshbirodkar**
See new figures
figure_1
figure_2
figure_3

### CODE REVIEW

**Comment by vighneshbirodkar** [comment]
@jni @JDWarner Thoughts ?

**Comment by jni** [comment]
@vighneshbirodkar this still could use some tweaking... I am not yet sure how, but these aren't very satisfying...

Here's my thinking: ultimately you want to show both the background image and the (continuous-valued) edge map. This is not possible when using maps like viridis that span the full color range. Instead, we should display the image using a full-range, monochromatic colormap, e.g. greens, and then the edges using a complementary colormap, e.g. something ranging along magenta and orange, that reaches neither black nor white in luminance.

Those were just examples — I don't know what colormaps we should choose, but that's what we should be aiming for imho.

**Comment by vighneshbirodkar** [comment]
Or could we dim the image ?

**Comment by jni** [comment]
@vighneshbirodkar the problem with dimming the image is that it becomes very difficult to see. The graph by itself is not interesting. What's interesting is the graph superimposed on the image.

**Comment by stefanv** [comment]
Or gray + any colormap?

**Comment by vighneshbirodkar** [comment]
@stefanv We already support that with the `desaturate` option

**Comment by JDWarner** [comment]
Personally, I would be fine with desaturated+color graph overlay as the default and possibly only method.

**Comment by vighneshbirodkar** [comment]
Line thickness helps a lot, here are some examples. 

## afmhot
afmhot

## inferno
inferno

## jet
jet

## magma
magma

## plasma 
plasma

## viridis
viridis

The script I used is here
https://gist.github.com/vighneshbirodkar/ca5908d513ec7993893a

**Comment by jni** [comment]
@vighneshbirodkar @JDWarner @stefanv The problem with the gray cmap is that bright areas and dark areas have low contrast with bright edges / dark edges on most colormaps. (The pixel boundaries don't help, either.)

**Comment by vighneshbirodkar** [comment]
@jni can you suggest a background color map and edge color map combination ? I chose to keep the boundaries because they demarcate the region.

**Comment by jni** [comment]
@vighneshbirodkar I agree the boundaries are useful. Maybe "mode='subpixel'" would be useful, although you would then have to transform all the other coordinates. Might be worthwhile. Another idea would be to draw the boundaries with some alpha.

For colormaps, obviously have a look [here](http://matplotlib.org/users/colormaps.html) for inspiration. After looking at that for a bit, I think bone (image) with cool (edges) might work quite well, actually! Also gist_heat with winter.

Guidelines for what we should be looking for:

- image itself should span the full luminance range
- graph edges should span a restricted range (avoiding the black and white ends of the spectrum) and use contrasting colors to the image.
- avoid red/green contrast. More colorblind friendly contrasts include blue-orange and green-purple.

Another suggestion would be to use a panel of a few more test images, since we might be overfitting to this coffee cup. =) But, having said that, I think you started with the most difficult example, since the image is very contrasty.

As a complete aside I'd like to give another shout-out to @stefanv and @njsmith for their stellar work on the new MPL colormaps. I keep hunting for their design page to send to various people sending me jet figures, and every time I find it I'm wowed by how thorough it is. Thanks!

**Comment by vighneshbirodkar** [comment]
### bone with cool
bone_cool


### gist_heat with winter
gist_heat_winter

I used an alpha of `.75` for all the edges this time

**Comment by jni** [comment]
Hmm, heat/winter doesn't work at all, but bone/cool looks pretty nice to me! What do you think?

**Comment by vighneshbirodkar** [comment]
I liked the plasma/viridis/magma over gray plots more.

**Comment by jni** [comment]
@vighneshbirodkar I think they are slightly more pleasing to the eye, but we are primarily concerned about a clear representation of the data. Have a look at the top-right corner of the image in those examples: I can't tell the difference between graph edges and fragment boundaries. I think that's a major problem. Similarly, with the top-left and bottom-center, the black of the image becomes muddled with edges and boundaries. It all just looks black. (Plasma is perhaps the least offensive in this regard.) Note that, with SLIC, our brains can complete the grid quite effectively. When we start using more irregular regions, it's going to be more difficult, and clarity of representation will be even more important.

@stefanv @JDWarner any further opinions?

**Comment by vighneshbirodkar** [comment]
@stefanv @JDWarner Can you give us your thoughts ? We can make the most popular choice the default.

**Comment by stefanv** [comment]
bone + cool looks great to me.  Have you tried bone + magma or bone +
viridis?  I think the above plots were gray + cmap.

**Comment by JDWarner** [comment]
@vighneshbirodkar I agree with @stefanv that bone for the underlying image seems to work well. A couple more comparison plots on top of bone and I think we can all be satisfied.

**Comment by vighneshbirodkar** [comment]
## bone + afmhot
afmhot

## bone + inferno
inferno

## bone + magma
magma

## bone + plasma
plasma

## bone + viridis
viridis

**Comment by vighneshbirodkar** [comment]
/ping @jni @JDWarner @stefanv 
Could you please take a quick look ?

**Comment by jni** [comment]
I like bone + inferno or magma

**Comment by JDWarner** [comment]
Bone + inferno for me. Accentuates the difference a little more. 

I could go either way on the naming. I see your point, but `show` reminds one of `plt.show()` from the matplotlib perspective.

**Comment by jni** [comment]
I'm happy to keep the name as `draw_rag`. I think there's a way to extract the image from mpl though...?

**Comment by stefanv** [comment]
I quite like bone + plasma, as it highlights the grid.  But I am definitely
not going to bike shed this, so please go ahead with whichever option you
prefer!

**Comment by vighneshbirodkar** [comment]
@jni We can get the image from mpl but do we will have to do `imshow` and that might interfere with
what the user is already plotting.

**Comment by jni** [comment]
@vighneshbirodkar not if you create a new figure with a different backend... ;) Even better might be to figure out whatever mpl is doing to colormap an image... That shouldn't be too difficult... (famous last words)

**Comment by vighneshbirodkar** [comment]
@jni 
Does `img_to_colormap` sound like a useful function to you ?

**Comment by jni** [comment]
@vighneshbirodkar Not at all! =D I would call it `colormap_image` because I like verbs, though. =)

**Comment by stefanv** [comment]
``colormap_image`` ++!

**Comment by vighneshbirodkar** [comment]
@jni @stefanv @JDWarner  The colormap PR reached a very anticlimactic end. But nonetheless we have a solution for rendering images with colromaps. However, we are still left with the issue of drawing thick lines.

**Comment by jni** [comment]
@vighneshbirodkar what issue is that? I suggest passing a "linewidth" keyword argument that defaults to whatever you think looks good? =)

**Comment by vighneshbirodkar** [comment]
@jni The prbolem is we cannot return an image. Unless we use matplotlib to draw and then get the image from the canvas. Although before doing anything I would like to consult @tacaswell

**Comment by jni** [comment]
@vighneshbirodkar what do you mean we "cannot return an image"? Are you saying draw_line doesn't support varying line thickness?

At any rate, I think line thickness can wait for a later PR, I would think. However, consulting @tacaswell does seem to be a wise decision at this point. =P

**Comment by tacaswell** [comment]
I am super confused what is going on here.  I had seen these images going by and assumed the lines were drawn with mpl `LineCollection` + `scatter` (both of which will take care of the underlying color mapping for you), but I see that you are using skimage to draw the lines.

What is the goal here?  To get nice figures assuming you have mpl, or to get a reasonable png out no matter what?  

If it is the first, I think delegating a bunch of this work to mpl (and taking an `Axes` as an optional input) would make more sense.  By letting Agg do the rasterization you would get better behavior on zooming and by using mpl artist for the edges / centers you could do things like interactively re-map the colors or hook into the `mouse_event` / `picker` framework to add in interactivity (click on a node, get it's numeric weight).  It would also allow you to output the edges as vector graphics.

**Comment by vighneshbirodkar** [comment]
@tacaswell If we change the behavior of `draw_rag` to draw with `matplotlib` itself, we can potentially mess up what the user is already trying to draw. Also, the user will need to call `plt.show()` to see the results. 

@jni `draw.line` does not support thickness. The plots with 1 pixel wide lines don't look nearly as good.

**Comment by vighneshbirodkar** [comment]
@tacaswell Just to clarify, the plots you see were generated by separate gist, instead of the `draw_rag` function
https://gist.github.com/vighneshbirodkar/ca5908d513ec7993893a

**Comment by stefanv** [comment]
How about using anti-aliased lines?
​

**Comment by jni** [comment]
@stefanv @vighneshbirodkar I'm inclined to agree with @tacaswell. What else will the user really want to draw on top of this, and why wouldn't they use mpl for it as well? At any rate, since it's in future, we have free reign to fiddle with the API.

**Comment by vighneshbirodkar** [comment]
@jni What I mean is if the user is already drawing something we might end up drawing on top of it. Also, I would suggest renaming this to `show_rag`

**Comment by stefanv** [comment]
I am happy to use mpl.
​

**Comment by tacaswell** [comment]
If you are concerned about messing with other things the user is doing plotting wise I suggest taking `ax` as an optional input and then doing

```python
if ax is None:
    fig, ax = plt.subplots()
```

That way if the user wants to use this in conjunction with other plotting they can pass in an axes to put it on (and if it messes with what ever else is there, that is their problem) or if the user does not pass in an `Axes` makes a new figure which will not (directly) mess with the user (it will do so _indirectly_ if the user is running other code that use `plt.plot` and such to plot to the current axes as the current axes will change via `plt.subplots()`.  If you are paranoid you can set the the current axes back to what it was before

```python
if ax is None:
    old_ax = plt.gca()  # which make create a figure 
    fig, ax = plt.subplots()
    plt.sca(old_ax)
```
but that runs the risk of creating two figures where you only needed one, but such are the risks of using global state).

**Comment by JDWarner** [comment]
:+1: for @tacaswell's suggestion. Use mpl across the board; allow a user to the option to pass their own axis to draw on.

**Comment by vighneshbirodkar** [comment]
@JDWarner @jni @stefanv 
Please have a look. Here are the new plots. I have removed the `node_color` attribute. Also, I have removed the ability to draw edges using a single color.

### plot_rag_draw.py
plot_rag_draw_1
plot_rag_draw_2

### plot_rag_boundary.py
plot_rag_boundary_1

For reference, here are their old plots
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_draw.html#example-segmentation-plot-rag-draw-py
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_boundary.html

**Comment by vighneshbirodkar** [comment]
The colormap magma is not preset in older versions of matplotlib. Should I add it like viridis ?

**Comment by JDWarner** [comment]
@vighneshbirodkar IMO, just add it. We probably should ship all four for backwards compatibility in future examples.

**Comment by jni** [comment]
I like them! I still think the boundaries are a bit too thick, but that might be inevitable.

**Comment by vighneshbirodkar** [comment]
@jni Is that a +1 from you ?

**Comment by soupault** [comment]
xref https://github.com/scikit-image/scikit-image/issues/1857

**Comment by vighneshbirodkar** [comment]
@jni @stefanv @JDWarner 
Please take a look at the new plots

plot_rag_boundary_1
plot_rag_draw_1
plot_rag_draw_2

**Comment by vighneshbirodkar** [comment]
I would be glad if someone could review this. Thanks to everyone so far for their patience

File: doc/examples/segmentation/plot_rag_draw.py [hunk 1]
```diff
@@ -8,33 +8,22 @@
...
-# http://www.colorcombos.com/color-schemes/2/ColorCombo2.html
-cmap = colors.ListedColormap(['#6599FF', '#ff9900'])
-out = graph.draw_rag(labels, g, img, node_color="#ffde00", colormap=cmap,
-                     thresh=30, desaturate=True)
-plt.figure()
-plt.title("RAG with edge weights less than 30, color "
-          "mapped between blue and orange.")
-plt.imshow(out)
+fig, ax = plt.subplots()
+ax.set_title('RAG drawn will all edges')
```
Comment (jni) [inline doc/examples/segmentation/plot_rag_draw.py]: will -> with

File: doc/examples/segmentation/plot_rag_draw.py [hunk 2]
```diff
@@ -8,33 +8,22 @@
...
+smap = graph.show_rag(labels, g, img, ax=ax)
+plt.colorbar(smap, fraction=0.03)
```
Comment (jni) [inline doc/examples/segmentation/plot_rag_draw.py]: what is the "fraction" kwarg? Please add a comment to specify what it means.

File: skimage/future/graph/rag.py [hunk 3]
```diff
@@ -413,14 +410,14 @@ def rag_boundary(labels, edge_map, connectivity=2):
     return rag
 
 
-def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
-             edge_color='#00ff00', colormap=None, thresh=np.inf,
-             desaturate=False, in_place=True):
+def show_rag(labels, rag, img, border_color='black', edge_width=1.5,
+             edge_cmap='magma', img_cmap='bone', thresh=np.inf, in_place=True,
+             ax=None):
```
Comment (jni) [inline skimage/future/graph/rag.py]: Comment for future reference, you don't need to tackle in this PR, but raise an issue in that case: can borders and edges have alpha? That might make it even more visible.

File: skimage/future/graph/rag.py [hunk 4]
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
     border_color : colorspec, optional
-        Any matplotlib colorspec.
-    node_color : colorspec, optional
-        Any matplotlib colorspec. Yellow by default.
-    edge_color : colorspec, optional
-        Any matplotlib colorspec. Green by default.
-    colormap : colormap, optional
-        Any matplotlib colormap. If specified the edges are colormapped with
-        the specified color map.
+        Color with which the borders of regions are drawn.
```
Comment (jni) [inline skimage/future/graph/rag.py]: of -> between would be clearer imho

File: skimage/future/graph/rag.py [hunk 5]
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+    edge_width : float, optional
+        The thickness with which the RAG edges are drawn.
+    edge_cmap : colormap, optional
+        Any matplotlib colormap with which the edges are drawn.
+    img_cmap : colormap, optional
+        Any matplotlib colormap with which the image is draw. If set to `None`
+        the image is drawn as it is.
     thresh : float, optional
-        Edges with weight below `thresh` are not drawn, or considered for color
+        Edges with weight above `thresh` are not drawn, or considered for color
```
Comment (jni) [inline skimage/future/graph/rag.py]: Hmm, this change makes me wonder... Some RAGs use distance as a metric (in which case this is probably the right thing to do), but others might use similarity, in which case this is the *wrong* thing to do. What do you think about "min_edge_value" and "max_edge_value"?

File: skimage/future/graph/rag.py [hunk 6]
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
-        Convert the image to grayscale before displaying. Particularly helps
-        visualization when using the `colormap` option.
     in_place : bool, optional
         If set, the RAG is modified in place. For each node `n` the function
         will set a new attribute ``rag.node[n]['centroid']``.
+    ax : matplotlib axes, optional
+        The axes to draw on.
 
-    Returns
+    Reutrns
```
Comment (jni) [inline skimage/future/graph/rag.py]: LOL Worst. Change. Ever. =P

File: skimage/future/graph/rag.py [hunk 7]
```diff
@@ -431,28 +428,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
     -------
-    out : ndarray, shape (M, N, 3)
-        The image with the RAG drawn.
+    smap : matplotlib ScalarMappable
+         A ScalarMappable for the edges that can be used to show colorbars.
```
Comment (jni) [inline skimage/future/graph/rag.py]: I have no idea what a ScalarMappable is, so perhaps adding a hyperlink reference to this would be useful.

Comment (jni) [inline skimage/future/graph/rag.py]: Actually I just saw the example below. Maybe say, "that can be passed to the `pyplot.colorbar()` to show the scale of the colormap used.

File: skimage/future/graph/rag.py [hunk 8]
```diff
@@ -461,20 +457,32 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+
     if not in_place:
         rag = rag.copy()
 
-    if desaturate:
-        img = color.rgb2gray(img)
-        img = color.gray2rgb(img)
-
+    if ax is None:
+        fix, ax = plt.subplots()
```
Comment (jni) [inline skimage/future/graph/rag.py]: fix -> fig. ie you should fix `fix`. =P

File: skimage/future/graph/rag.py (hunk 7 above)
Comment (tacaswell) [inline skimage/future/graph/rag.py]: There is magic you can use with Sphinx that will auto-link to the mpl docs (`intersphinx_mapping` is the key in  `confg.py`)

File: skimage/future/graph/rag.py [hunk 9]
```diff
@@ -491,31 +499,27 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
-        circle = draw.circle(r1, c1, 2)
-        out[circle] = node_color
+    for n1, n2, data in rag.edges_iter(data=True):
 
-    return out
+        if data['weight'] < thresh:
+            r1, c1 = map(int, rag.node[n1]['centroid'])
+            r2, c2 = map(int, rag.node[n2]['centroid'])
+            col = smap.to_rgba([data['weight']])[0][:-1]
+            ax.add_artist(plt.Line2D([c1, c2], [r1, r2],  color=col,
```
Comment (tacaswell) [inline skimage/future/graph/rag.py]: I would use a `LineCollection` for this, you get better prefromance and mpl will take care of the norm/color mapping for you.

File: skimage/future/graph/rag.py [hunk 10]
```diff
@@ -461,20 +457,32 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+        fix, ax = plt.subplots()
     out = util.img_as_float(img, force_copy=True)
-    cc = colors.ColorConverter()
 
-    edge_color = cc.to_rgb(edge_color)
-    node_color = cc.to_rgb(node_color)
+    if img_cmap is None:
+        if img.ndim < 3:
+            raise ValueError('The image needs to be RGB')
+        if img.ndim < 4:
```
Comment (jni) [inline skimage/future/graph/rag.py]: What's all this? It seems brittle. If I give you a 3D grayscale image, you'll give me the first three columns blended as RGB and ignore the rest.

On the other hand, while we're assuming things, why not assume grayscale when the image is 2D and they've used colormap=None? Or, at least, perhaps the ValueError should specify that it should be RGB *or* a colormap should be selected.

File: skimage/future/graph/rag.py [hunk 11]
```diff
@@ -461,20 +457,32 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+            new_img = np.empty((r, c, 4), dtype=np.float)
+            new_img[:, :, 0:3] = img
+            new_img[:, :, 3] = 1
+            out = new_img
+    else:
+        img_cmap = cm.get_cmap(img_cmap)
+        out = color.rgb2gray(img)
+        out = img_cmap(out)
+
+    edge_cmap = cm.get_cmap(edge_cmap)
```
Comment (jni) [inline skimage/future/graph/rag.py]: I think we've discussed this, but can you remind me again, why does this work when `edge_cmap` is `'viridis'` and the matplotlib version is < 1.5?

Comment (jni) [inline skimage/future/graph/rag.py]: Nevermind, just found it below, thank you! =)

**Comment by jni** [comment]
@vighneshbirodkar I've made some minor comments. Other than those, this is good, I'll be happy to (finally) merge! Thank you!

File: skimage/future/graph/rag.py (hunk 5 above)
Comment (vighneshbirodkar) [inline skimage/future/graph/rag.py]: I'm not a fan of threshold in the first place. If someone wants to omit some edges they can always remove them from the graph.

And I don't think there is a right or wrong thing to do for any kind of graph, it really depends on what the user is trying to do.

File: skimage/future/graph/rag.py (hunk 6 above)
Comment (vighneshbirodkar) [inline skimage/future/graph/rag.py]: I don't know how I managed to do this but I laughed for the better part of a minute reading your comment.

File: skimage/future/graph/rag.py (hunk 5 above)
Comment (jni) [inline skimage/future/graph/rag.py]: It's totally fine to remove the threshold parameter altogether as far as I'm concerned!

File: skimage/future/graph/rag.py [hunk 12]
```diff
@@ -461,20 +456,30 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
-        img = color.gray2rgb(img)
-
+    if ax is None:
+        fig, ax = plt.subplots()
     out = util.img_as_float(img, force_copy=True)
-    cc = colors.ColorConverter()
 
-    edge_color = cc.to_rgb(edge_color)
-    node_color = cc.to_rgb(node_color)
+    if img_cmap is None:
```
Comment (vighneshbirodkar) [inline skimage/future/graph/rag.py]: @jni This seems ok to you ?

**Comment by vighneshbirodkar** [comment]
The 2nd plot in `plot_rag_draw.py` looks like this
plot_rag_draw_2

File: skimage/future/graph/rag.py (hunk 12 above)
Comment (jni) [inline skimage/future/graph/rag.py]: @vighneshbirodkar I would additionally check that `img.shape[2] <= 4`

Comment (vighneshbirodkar) [inline skimage/future/graph/rag.py]: @jni Why not exactly equal to 3 ?

Comment (vighneshbirodkar) [inline skimage/future/graph/rag.py]: I would do `3 <= img.shape[2] <= 4`

Comment (soupault) [inline skimage/future/graph/rag.py]: `img.shape[2] in [3, 4]`?

Comment (jni) [inline skimage/future/graph/rag.py]: I believe @soupault's version to be both the fastest, most concise, and most standard (ie that pattern exists elsewhere in skimage)

**Comment by vighneshbirodkar** [comment]
@jni @soupault If you can have another look. It shouldn't be far from merge now

File: skimage/future/graph/rag.py [hunk 13]
```diff
@@ -506,6 +506,7 @@ def show_rag(labels, rag, img, border_color='black', edge_width=1.5,
     # Defining the end points of the edges
     # The tuple[::-1] syntax reverses a tuple as matplotlib uses (x,y)
     # convention while skimage uses (row, column)
+    ll_ = [(rag.node[n1]['centroid'], rag.node[n2]['centroid']) for (n1, n2) in rag.edges_iter()]
```
Comment (jni) [inline skimage/future/graph/rag.py]: This goes unused?

Comment (soupault) [inline skimage/future/graph/rag.py]: @jni https://github.com/scikit-image/scikit-image/pull/1872/files#diff-ad47b78e529d8acd83e9c6bbb84874bbR513

File: skimage/future/graph/rag.py [hunk 14]
```diff
@@ -429,30 +426,28 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+    edge_cmap : colormap, optional
+        Any matplotlib colormap with which the edges are drawn.
+    img_cmap : colormap, optional
+        Any matplotlib colormap with which the image is draw. If set to `None`
+        the image is drawn as it is.
     in_place : bool, optional
         If set, the RAG is modified in place. For each node `n` the function
         will set a new attribute ``rag.node[n]['centroid']``.
+    ax : matplotlib axes, optional
+        The axes to draw on.
```
Comment (soupault) [inline skimage/future/graph/rag.py]: I'm reviewing this for the first time and this confused me.
Probably, as far as `ax` is `optonal`, it's worth to add a note on default behaviour.

File: skimage/future/graph/rag.py [hunk 15]
```diff
@@ -489,33 +494,25 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+    # Defining the end points of the edges
+    # The tuple[::-1] syntax reverses a tuple as matplotlib uses (x,y)
+    # convention while skimage uses (row, column)
+    ll_ = [(rag.node[n1]['centroid'], rag.node[n2]['centroid']) for (n1, n2) in rag.edges_iter()]
+    lines = [[rag.node[n1]['centroid'][::-1], rag.node[n2]['centroid'][::-1]]
+             for (n1, n2) in rag.edges_iter()]
 
-        circle = draw.circle(r1, c1, 2)
-        out[circle] = node_color
+    lc = LineCollection(lines, linewidths=edge_width, cmap=edge_cmap)
```
Comment (vighneshbirodkar) [inline skimage/future/graph/rag.py]: @jni. LineCollection is used here

**Comment by jni** [comment]
@vighneshbirodkar and `ll_`?

**Comment by vighneshbirodkar** [comment]
@jni Fixed that

File: skimage/future/graph/rag.py [hunk 16]
```diff
@@ -429,30 +426,28 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
     rag : RAG
         The Region Adjacency Graph.
     img : ndarray, shape (M, N, 3)
```
Comment (jni) [inline skimage/future/graph/rag.py]: shape (M, N[, 3]), right?

File: skimage/future/graph/rag.py [hunk 17]
```diff
@@ -429,30 +426,28 @@ def draw_rag(labels, rag, img, border_color=None, node_color='#ffff00',
...
+    ax : matplotlib axes, optional
+        The axes to draw on.
 
     Returns
     -------
-    out : ndarray, shape (M, N, 3)
-        The image with the RAG drawn.
+    lc : :py:class:`matplotlib.collections.LineCollection`
+         A colection of lines that represent the edges of the graph. It can be
+         passed to the :meth:`matplotlib.figure.Figure.colorbar` function.
```
Comment (jni) [inline skimage/future/graph/rag.py]: :100:

File: skimage/util/colormap.py [hunk 18]
```diff
@@ -258,4 +259,266 @@
...
+              [0.988717, 0.955742, 0.712242],
+              [0.988367, 0.962878, 0.719649],
+              [0.988033, 0.970012, 0.727077],
+              [0.987691, 0.977154, 0.734536],
+              [0.987387, 0.984288, 0.742002],
+              [0.987053, 0.991438, 0.749504]]
+
+
 viridis = LinearSegmentedColormap.from_list('viridis', viridis_data)
+cm.register_cmap('viridis', viridis)
```
Comment (jni) [inline skimage/util/colormap.py]: I would do this only if they are not already registered, so as not to overwrite whatever ships with matplotlib >1.5, which might have some improvements over what we ship here, for example.

**Comment by jni** [comment]
@vighneshbirodkar this is almost done! Just two very minor comments.

**Comment by vighneshbirodkar** [comment]
@jni @soupault All tests pass now. Please have look

**Comment by vighneshbirodkar** [comment]
Looks like the AppVeyor build timed out!

**Comment by jni** [comment]
@vighneshbirodkar it's been doing that. No big deal. =)

This gets a 👍 from me! Thanks for your infinite patience, @vighneshbirodkar! =)

**Comment by soupault** [comment]
@vighneshbirodkar have you tried to supplement these new plots with RAG vertices? Can't see any point in the discussion where we decided to drop them.
The PR looks good to me, but I'd like to clarify this first.

**Comment by jni** [comment]
@soupault I would say that vertices never really added much to the drawing, and we spent quite enough time bikeshedding on this...

**Comment by vighneshbirodkar** [comment]
@soupault

I mentioned that I removed the node_color attribute.@JDWarner once pointed that out and even I feel that node color and is one extra parameter the user has to choose while it does nothing to improve the drawing. That's also the reason I removed the ability to draw all edges with a single color.

And I feel this function should just get out of the user's way. It's not doing any work as such and it's only purpose is to provide insight on what to do next. The less time the user spends customising this, the better.

**Comment by soupault** [comment]
@jni @vighneshbirodkar thanks for your answers! Cannot see any obstacles to finally get this alive then.
Thank you for your patience, @vighneshbirodkar.
Merging.

**Comment by stefanv** [comment]
\o/

**Comment by soupault** [comment]
Everything regarding this PR is fine:
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_boundary.html
http://scikit-image.org/docs/dev/auto_examples/segmentation/plot_rag_draw.html

But, Google has [this page indexed](http://scikit-image.org/docs/dev/auto_examples/plot_rag_draw.html), which is actually a mess: the plotted images do not correspond to the code. Can we do anything about that? /cc @emmanuelle

**Comment by vighneshbirodkar** [comment]
@soupault Can't we just delete that page ?

**Comment by soupault** [comment]
Only manually, I suppose. I don't have these permissions.
Btw, not just this page should be removed, but also the pages for another examples ([1](http://scikit-image.org/docs/dev/auto_examples/plot_regional_maxima.html), [2](http://scikit-image.org/docs/dev/auto_examples/plot_equalize.html), etc). This should have been accounted in https://github.com/scikit-image/scikit-image/pull/1823 :/.

**Comment by stefanv** [comment]
Our doc upload script should be updated to not only update and upload new files but to also remove old files.  @soupault I'll give you permissions on that repo as well.

**Pull request merged**